test/randomize_order_test.py failed

Error generated:
```
=================================== FAILURES ===================================
___________________________ TestRandomizeDipnet.test ___________________________

self = <randomize_order_test.TestRandomizeDipnet object at 0x7fa0753b7c88>

    def test(self):
    
        # The following three tests check that build, disband and remove commands do not get changed when input into the order randomizer.
    
        ord = [(("FRA", "AMY", "PAR"), "BLD")]
        assert ord == random_list_orders(ord)
    
        ord = [(("FRA", "AMY", "PAR"), "DSB")]
        assert ord == random_list_orders(ord)
    
        ord = [(("FRA", "AMY", "PAR"), "REM")]
        assert ord == random_list_orders(ord)
    
        # The following three tests check that when orders that contain movements, holds, convoys and other moves get
        # input into the order randomizer, they come out differnent.
    
        orders = [
            (("FRA", "FLT", "NTH"), "CVY", ("FRA", "AMY", "HOL"), "CTO", "NWY"),
            (("FRA", "AMY", "HOL"), "CTO", "NWY", "VIA", ("NTH")),
            (("FRA", "AMY", "BER"), "HLD"),
        ]
>       assert random_list_orders(orders) != orders

tests/randomize_order_test.py:35: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/baseline_bots/randomize_order.py:109: in random_list_orders
    map(lambda order: randomize(order), orders)
src/baseline_bots/randomize_order.py:109: in <lambda>
    map(lambda order: randomize(order), orders)
src/baseline_bots/randomize_order.py:168: in randomize
    return tag_to_func[tag](order)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

order = (('FRA', 'AMY', 'HOL'), 'CTO', 'NWY', 'VIA', 'NTH')

    def random_convoy_to(order: Tuple) -> Tuple:
        """
        This takes a convoy order and returns the longest alternate convoy.
    
        :param order: A "convoy to" (CTO) order
        :type order: Tuple
        :return: A deviant order (with some chance of being the same order).
        :rtype: Tuple
        """
        (_, _, amy_loc), _, province, _, (sea_provinces) = order
        sea_provinces = list(reversed(sea_provinces))
        for i, sea in enumerate(
            sea_provinces
        ):  # searches through the sea provinces in reversed order to find the longest possible alternate convoy
            # fmt : off
            valid = [
                loc
>               for loc in ADJACENCY[sea]
                if TYPES[loc] == "COAST"
                and loc != [province]
                and loc not in ADJACENCY[amy_loc]
            ]  # the location must not be the one the unit is already convoying to
E           KeyError: 'H'

src/baseline_bots/randomize_order.py:188: KeyError
=========================== short test summary info ============================
FAILED tests/randomize_order_test.py::TestRandomizeDipnet::test - KeyError: 'H'
============================== 1 failed in 0.40s ===============================
The command '/bin/sh -c pytest tests/randomize_order_test.py' returned a non-zero code: 1
Error: Process completed with exit code 1.
```

--- a/src/baseline_bots/bots/dipnet/transparent_bot.py
+++ b/src/baseline_bots/bots/dipnet/transparent_bot.py
@@ -53,7 +53,9 @@ class TransparentBot(DipnetBot):
             return comms_obj
         parsed_orders = self.parse_messages(rcvd_messages)
         parsed_orders = [
-            list(daide_to_dipnet_parsing(order))[0] for order in parsed_orders if daide_to_dipnet_parsing(order)
+            list(daide_to_dipnet_parsing(order))[0]
+            for order in parsed_orders
+            if daide_to_dipnet_parsing(order)
         ]
 
         # My orders' messages if not already sent

--- a/src/baseline_bots/bots/random_proposer_bot.py
+++ b/src/baseline_bots/bots/random_proposer_bot.py
@@ -94,13 +94,13 @@ class RandomProposerBot_AsyncBot(RandomProposerBot):
         :return: dict containing messages and orders
         """
         messages = yield self.gen_messages(rcvd_messages)
-        
+
         if messages and messages.messages:
             for msg in messages.messages:
                 msg_obj = Message(
                     sender=self.power_name,
-                    recipient=msg['recipient'],
-                    message=msg['message'],
+                    recipient=msg["recipient"],
+                    message=msg["message"],
                     phase=self.game.get_current_phase(),
                 )
                 yield self.game.send_game_message(message=msg_obj)

--- a/src/baseline_bots/bots/smart_order_accepter_bot.py
+++ b/src/baseline_bots/bots/smart_order_accepter_bot.py
@@ -5,13 +5,11 @@ import random
 from collections import defaultdict
 from typing import Dict, List, Set, Tuple
 
+import numpy as np
 from DAIDE import FCT, HUH, ORR, PRP, XDO, YES
 from diplomacy import Message
 from stance_vector import ActionBasedStance, ScoreBasedStance
-import random
-import numpy as np
 from tornado import gen
-
 
 from baseline_bots.bots.dipnet.dipnet_bot import DipnetBot
 from baseline_bots.parsing_utils import (
@@ -20,6 +18,7 @@ from baseline_bots.parsing_utils import (
     parse_proposal_messages,
 )
 from baseline_bots.randomize_order import (
+    lst_to_daide,
     random_list_orders,
     string_to_tuple,
     tuple_to_string,
@@ -33,7 +32,7 @@ from baseline_bots.utils import (
     get_other_powers,
     smart_select_support_proposals,
 )
-from baseline_bots.randomize_order import lst_to_daide
+
 
 class SmartOrderAccepterBot(DipnetBot):
     """
@@ -52,7 +51,9 @@ class SmartOrderAccepterBot(DipnetBot):
         super().__init__(power_name, game)
         self.alliance_props_sent = False
         self.discount_factor = discount_factor
-        self.stance = ActionBasedStance(power_name, game, discount_factor=self.discount_factor)
+        self.stance = ActionBasedStance(
+            power_name, game, discount_factor=self.discount_factor
+        )
         self.alliances = defaultdict(list)
         self.rollout_length = 5
         self.rollout_n_order = 5
@@ -64,7 +65,7 @@ class SmartOrderAccepterBot(DipnetBot):
         self.allies = []
         self.foes = []
         self.neutral = []
-        
+
     async def send_message(self, recipient: str, message: MessagesData) -> None:
         """
         Send message asynchronously to the server while the bot is still processing
@@ -101,15 +102,15 @@ class SmartOrderAccepterBot(DipnetBot):
             )
             if str(orders_decided) != "FCT ()":
                 for pow in self.allies:
-                    if (
-                        pow != self.power_name
-                    ):
+                    if pow != self.power_name:
                         msgs_data.add_message(pow, str(orders_decided))
                         await self.send_message(pow, str(orders_decided))
 
-    async def gen_messages(self, orders_list: List[str], msgs_data: MessagesData) -> MessagesData:
+    async def gen_messages(
+        self, orders_list: List[str], msgs_data: MessagesData
+    ) -> MessagesData:
         """
-        This generates messages to be sent to the other powers. 
+        This generates messages to be sent to the other powers.
         Note: Messages are also generated outside of this function invocation flow
 
         :param orders_list: final list of orders decided by the bot
@@ -132,10 +133,7 @@ class SmartOrderAccepterBot(DipnetBot):
         :return: MessagesData object containing set of all messages
         """
         for proposer, orders in prp_orders.items():
-            if (
-                orders
-                and self.power_name != proposer
-            ):
+            if orders and self.power_name != proposer:
                 if proposer == best_proposer and proposer in self.allies:
                     msg = YES(
                         PRP(
@@ -353,7 +351,9 @@ class SmartOrderAccepterBot(DipnetBot):
             provinces.update(new_provs)
         return provinces
 
-    async def generate_support_proposals(self, comms_obj: MessagesData) -> Dict[str, str]:
+    async def generate_support_proposals(
+        self, comms_obj: MessagesData
+    ) -> Dict[str, str]:
         """
         Using the orders already decided, search the neighbourhood provinces for allies' units and generate support proposals accordingly
 
@@ -428,9 +428,9 @@ class SmartOrderAccepterBot(DipnetBot):
                 final_messages[recipient] = str(suggested_proposals)
                 comms_obj.add_message(recipient, str(suggested_proposals))
                 await self.send_message(recipient, str(suggested_proposals))
-                
+
         return final_messages
-        
+
     def is_order_aggressive_to_powers(self, order: str, powers: List[str]):
         """
         check if the order is aggressive by
@@ -487,23 +487,34 @@ class SmartOrderAccepterBot(DipnetBot):
         order_token = get_order_tokens(order)
         unit = order_token[0]
         loc_unit = unit[2:]
-        list_order, prob_order = yield self.brain.get_beam_orders(self.game, self.power_name)
-    
-        if len(list_order)>1:
-            for i in range(1,len(list_order)):
+        list_order, prob_order = yield self.brain.get_beam_orders(
+            self.game, self.power_name
+        )
+
+        if len(list_order) > 1:
+            for i in range(1, len(list_order)):
                 dipnet_order = list_order[i]
                 for candidate_order in dipnet_order:
-                    if unit in candidate_order and not self.is_order_aggressive_to_powers(candidate_order, powers):
+                    if (
+                        unit in candidate_order
+                        and not self.is_order_aggressive_to_powers(
+                            candidate_order, powers
+                        )
+                    ):
                         return candidate_order
-        
+
         # if none in dipnet beam orders
         for current_order in self.orders.get_list_of_orders():
-            if current_order != order and not self.is_order_aggressive_to_powers(current_order, powers) and current_order in self.game.get_all_possible_orders()[loc_unit]:
-                return unit + ' S ' + current_order
-        
-        return unit + ' H'
+            if (
+                current_order != order
+                and not self.is_order_aggressive_to_powers(current_order, powers)
+                and current_order in self.game.get_all_possible_orders()[loc_unit]
+            ):
+                return unit + " S " + current_order
 
-    @gen.coroutine                   
+        return unit + " H"
+
+    @gen.coroutine
     def replace_aggressive_order_to_allies(self):
         """
         replace aggressive orders with non-aggressive orders (in-place replace self.orders)
@@ -521,22 +532,21 @@ class SmartOrderAccepterBot(DipnetBot):
             else:
                 new_order = order
             final_orders.append(new_order)
-        
+
         orders_data = OrdersData()
         orders_data.add_orders(final_orders)
         self.orders = orders_data
-
 
     @gen.coroutine
     def __call__(self, rcvd_messages: List[Tuple[int, Message]]):
         # compute pos/neg stance on other bots using Tony's stance vector
 
         # avoid get_stance in the first phase of game
-        if self.game.get_current_phase()!='S1901M':
+        if self.game.get_current_phase() != "S1901M":
             self.stance.get_stance(self.game)
         print(f"Stance vector for {self.power_name}")
         print(self.stance.stance[self.power_name])
-            
+
         powers = self.stance.stance[self.power_name]
 
         # get dipnet order
@@ -549,7 +559,7 @@ class SmartOrderAccepterBot(DipnetBot):
         print("debug: Fetched orders", orders)
 
         # only in movement phase, we send PRP/ALY/FCT and consider get_best_proposer
-        if self.game.get_current_phase()[-1]=='M':
+        if self.game.get_current_phase()[-1] == "M":
 
             # parse the proposal messages received by the bot
             parsed_messages_dict = parse_proposal_messages(
@@ -566,9 +576,6 @@ class SmartOrderAccepterBot(DipnetBot):
             # e.g. if we are playing as ENG and the base orders are generated from DipNet, we would want to consider
             # if there is any better proposal orders that has a state value more than ours, then do it. If not, just follow the base orders.
             valid_proposal_orders[self.power_name] = orders
-
-
-
 
             # fmt: off
 

--- a/src/baseline_bots/parsing_utils.py
+++ b/src/baseline_bots/parsing_utils.py
@@ -66,8 +66,7 @@ def dipnet_to_daide_parsing(
                 f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing: unit {dipnet_suborder} not present in unit_game_mapping"
             )
             return None
-        
-        
+
         unit_type = "AMY" if dipnet_suborder[0] == "A" else "FLT"
         unit = expand_prov_coast(dipnet_suborder.split()[-1])
 
@@ -144,7 +143,9 @@ def dipnet_to_daide_parsing(
             if daidefy_suborder(dipnet_order_tokens[0]):
                 daide_order.append(daidefy_suborder(dipnet_order_tokens[0]))
             else:
-                print(f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing() skipping message: {dipnet_order_tokens}")
+                print(
+                    f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing() skipping message: {dipnet_order_tokens}"
+                )
                 continue
             if dipnet_order_tokens[1] == "S":
                 # Support orders
@@ -152,12 +153,16 @@ def dipnet_to_daide_parsing(
                 if daidefy_suborder(dipnet_order_tokens[2]):
                     daide_order.append(daidefy_suborder(dipnet_order_tokens[2]))
                 else:
-                    print(f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing() skipping message: {dipnet_order_tokens}")
+                    print(
+                        f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing() skipping message: {dipnet_order_tokens}"
+                    )
                     continue
 
                 if len(dipnet_order_tokens) == 4 and dipnet_order_tokens[3] != "H":
                     daide_order.append("MTO")
-                    daide_order.append(expand_prov_coast(dipnet_order_tokens[3].split()[-1]))
+                    daide_order.append(
+                        expand_prov_coast(dipnet_order_tokens[3].split()[-1])
+                    )
                 elif len(dipnet_order_tokens) > 4:
                     print(
                         f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing: order {dipnet_order_tokens} is UNEXPECTED. Update code to handle this case!!!"
@@ -172,15 +177,21 @@ def dipnet_to_daide_parsing(
                 if daidefy_suborder(dipnet_order_tokens[2]):
                     daide_order.append(daidefy_suborder(dipnet_order_tokens[2]))
                 else:
-                    print(f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing() skipping message: {dipnet_order_tokens}")
+                    print(
+                        f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing() skipping message: {dipnet_order_tokens}"
+                    )
                     continue
 
                 daide_order.append("CTO")
-                daide_order.append(expand_prov_coast(dipnet_order_tokens[3].split()[-1]))
+                daide_order.append(
+                    expand_prov_coast(dipnet_order_tokens[3].split()[-1])
+                )
             elif len(dipnet_order_tokens) >= 3 and dipnet_order_tokens[2] == "VIA":
                 # VIA/CTO orders
                 daide_order.append("CTO")
-                daide_order.append(expand_prov_coast(dipnet_order_tokens[1].split()[-1]))
+                daide_order.append(
+                    expand_prov_coast(dipnet_order_tokens[1].split()[-1])
+                )
                 daide_order.append("VIA")
                 if dipnet_order_tokens[0] + dipnet_order_tokens[1] in convoy_map:
                     daide_order.append(
@@ -194,7 +205,9 @@ def dipnet_to_daide_parsing(
             else:
                 # Move orders
                 daide_order.append("MTO")
-                daide_order.append(expand_prov_coast(dipnet_order_tokens[1].split()[-1]))
+                daide_order.append(
+                    expand_prov_coast(dipnet_order_tokens[1].split()[-1])
+                )
                 if len(dipnet_order_tokens) > 2:
                     print(
                         f"ALLAN: error from parsing_utils.dipnet_to_daide_parsing: order {dipnet_order_tokens} is UNEXPECTED. Update code to handle this case!!!"
@@ -271,7 +284,9 @@ def daide_to_dipnet_parsing(daide_style_order_str: str) -> Tuple[str, str]:
         try:
             ans = suborder_tokens[1][0] + " " + compress_prov_coast(suborder_tokens[2])
         except Exception:
-            print(f"ALLAN: error from parsing_utils.daide_to_dipnet_parsing.dipnetify_suborder() Failed for suborder: {suborder_tokens}")
+            print(
+                f"ALLAN: error from parsing_utils.daide_to_dipnet_parsing.dipnetify_suborder() Failed for suborder: {suborder_tokens}"
+            )
             ans = suborder_tokens[1][0] + " " + suborder_tokens[2]
         return ans, suborder_tokens[0]
 
@@ -287,7 +302,10 @@ def daide_to_dipnet_parsing(daide_style_order_str: str) -> Tuple[str, str]:
             # Support order
             dipnet_order.append("S")
             dipnet_order.append(dipnetify_suborder(daide_style_order_groups[2])[0])
-            if len(daide_style_order_groups) == 5 and daide_style_order_groups[3] == "MTO":
+            if (
+                len(daide_style_order_groups) == 5
+                and daide_style_order_groups[3] == "MTO"
+            ):
                 dipnet_order.append("-")
                 dipnet_order.append(compress_prov_coast(daide_style_order_groups[4]))
             elif len(daide_style_order_groups) > 5:
@@ -390,7 +408,9 @@ def parse_proposal_messages(
                     else:
                         other_orders[order_msg.sender].append(order)
             except Exception as e:
-                print(f"ALLAN: error from parsing_utils.parse_proposal_messages() Unexpected message format: {order_msg.message}")
+                print(
+                    f"ALLAN: error from parsing_utils.parse_proposal_messages() Unexpected message format: {order_msg.message}"
+                )
                 continue
 
         # Generate set of possible orders for the given power

--- a/src/baseline_bots/randomize_order.py
+++ b/src/baseline_bots/randomize_order.py
@@ -177,9 +177,11 @@ def random_convoy_to(order: Tuple) -> Tuple:
     :return: A deviant order (with some chance of being the same order).
     :rtype: Tuple
     """
-    (_, _, amy_loc), _, province, _, (sea_provinces) = order
-    sea_provinces = sea_provinces.split()
-    sea_provinces = list(reversed(sea_provinces))
+    (_, _, amy_loc), _, province, _, sea_provinces = order
+    if isinstance(sea_provinces, Tuple):
+        sea_provinces = list(reversed(sea_provinces))
+    else:
+        sea_provinces = [sea_provinces]
     for i, sea in enumerate(
         sea_provinces
     ):  # searches through the sea provinces in reversed order to find the longest possible alternate convoy
@@ -369,10 +371,11 @@ def string_to_tuple(orders: str) -> Tuple:
     )  # inserts quotes around strings
     return eval(with_quotes)
 
+
 def lst_to_daide(orders: List) -> str:
-    '''
+    """
     This function should take DAIDE orders as a list of strings and wrap them so: FCT ( ORR ( XDO(ORD1) XDO(ORD2) ) )
-    '''
+    """
     daide_ords = "FCT (ORR"
     for ord in orders:
         daide_ords += " (XDO (" + ord + "))"

--- a/src/baseline_bots/utils.py
+++ b/src/baseline_bots/utils.py
@@ -448,7 +448,7 @@ def get_best_orders(bot, proposal_order: dict, shared_order: dict):
         setattr(result, "powers", {})
         for power in game.powers.values():
             result.powers[power.name] = deepcopy(power)
-            setattr(result.powers[power.name], 'game', result)
+            setattr(result.powers[power.name], "game", result)
         result.role = strings.SERVER_TYPE
         return result
 

--- a/tests/randomize_order_test.py
+++ b/tests/randomize_order_test.py
@@ -51,7 +51,6 @@ class TestRandomizeDipnet:
 
         # This tests the function randomize_joiner which makes sure that the orders ouput are different than the orders input
         test_string = "AND ((FRA AMY BUR) MTO BEL) ((FRA AMY PIC) CTO FIN VIA (NTH SKA DEN BAL BOT))"
-        print(randomize_order(test_string))
         assert test_string != randomize_order(test_string)
 
         # These following tests test the validity of orders that the randomizer can generate

--- a/tests/randomize_order_test.py
+++ b/tests/randomize_order_test.py
@@ -32,11 +32,11 @@ class TestRandomizeDipnet:
 
         orders = [(("FRA", "FLT", "NTH"), "CVY", ('FRA', 'AMY', 'HOL'), 'CTO', "NWY"), (("FRA", "AMY", "HOL"), "CTO", 'NWY',"VIA", ('NTH'))]
         random.seed(1)
-        assert random_list_orders(orders) == [(('FRA', 'FLT', 'NTH'), 'MTO', 'DEN'), (('FRA', 'AMY', 'HOL'), 'MTO', 'NTH')]
+        assert random_list_orders(orders) == [(('FRA', 'FLT', 'NTH'), 'CVY', ('FRA', 'AMY', 'HOL'), 'CTO', 'EDI'), (('FRA', 'AMY', 'HOL'), 'CTO', 'NWY', 'VIA', ('NTH',))]
 
         orders = [(("FRA", "FLT", "NTH"), "CVY", ('FRA', 'AMY', 'HOL'), 'CTO', "NWY"), (("FRA", "AMY", "HOL"), "CTO", 'NWY',"VIA", ('NTH'))]
         random.seed(15)
-        assert random_list_orders(orders) ==[(('FRA', 'FLT', 'NTH'), 'HLD'), (('FRA', 'AMY', 'HOL'), 'MTO', 'BEL')]
+        assert random_list_orders(orders) ==[(('FRA', 'FLT', 'NTH'), 'CVY', ('FRA', 'AMY', 'HOL'), 'CTO', 'EDI'), (('FRA', 'AMY', 'HOL'), 'CTO', 'DEN', 'VIA', ('NTH',))]
         
         # This tests the ability for string_to_tuple to convert this string representing
         # a "convoy to" order properly
@@ -51,6 +51,7 @@ class TestRandomizeDipnet:
 
         # This tests the function randomize_joiner which makes sure that the orders ouput are different than the orders input
         test_string = "AND ((FRA AMY BUR) MTO BEL) ((FRA AMY PIC) CTO FIN VIA (NTH SKA DEN BAL BOT))"
+        print(randomize_order(test_string))
         assert test_string != randomize_order(test_string)
 
         # These following tests test the validity of orders that the randomizer can generate

